test/common/upstream/health_checker_impl_test.cc

```diff
@@ -1452,6 +1494,860 @@ TEST_F(RedisHealthCheckerImplTest, AllDontReuseConnection) {
+            {},
+            {},
+        },
+        // End stream on response headers.
+        {
+            {{":status", "200"}, {"content-type", "application/grpc"}},
+            {},
+            {},
+        },
+        // Non-OK gRPC status in headers.
+        {
+            {{":status", "200"}, {"content-type", "application/grpc"}, {"grpc-status", "2"}},
+            {},
+            {},
+        },
+        // Non-OK gRPC status
+        {
+            {{":status", "200"}, {"content-type", "application/grpc"}},
+            {GrpcHealthCheckerImplTest::ResponseSpec::servingResponse()},
+            {{"grpc-status", "2"}},
+        },
+        // Missing body.
+        {
+            {{":status", "200"}, {"content-type", "application/grpc"}, {"grpc-status", "0"}},
+            {},
+            {},
+        },
+        // Compressed body.
+        {
+            {{":status", "200"}, {"content-type", "application/grpc"}},
+            {GrpcHealthCheckerImplTest::ResponseSpec::invalidPayload(Grpc::GRPC_FH_COMPRESSED,
+                                                                     true)},
+            {},
+        },
+        // Invalid proto message.
+        {
+            {{":status", "200"}, {"content-type", "application/grpc"}},
+            {GrpcHealthCheckerImplTest::ResponseSpec::invalidPayload(Grpc::GRPC_FH_DEFAULT, false)},
+            {},
+        },
+        // Duplicate response.
+        {
+            {{":status", "200"}, {"content-type", "application/grpc"}},
+            {GrpcHealthCheckerImplTest::ResponseSpec::servingResponse(),
+             GrpcHealthCheckerImplTest::ResponseSpec::servingResponse()},
+            {},
+        },
+        // Invalid response.
+        {
+            {{":status", "200"}, {"content-type", "application/grpc"}},
+            {GrpcHealthCheckerImplTest::ResponseSpec::invalidChunk()},
+            {},
+        },
+        // No trailers.
+        {
+            {{":status", "200"}, {"content-type", "application/grpc"}},
+            {GrpcHealthCheckerImplTest::ResponseSpec::servingResponse()},
+            {},
+        },
+        // No gRPC status in trailer.
+        {
+            {{":status", "200"}, {"content-type", "application/grpc"}},
+            {GrpcHealthCheckerImplTest::ResponseSpec::servingResponse()},
+            {{"some-header", "1"}},
+        },
+        // Invalid gRPC status.
+        {
+            {{":status", "200"}, {"content-type", "application/grpc"}},
+            {GrpcHealthCheckerImplTest::ResponseSpec::servingResponse()},
+            {{"grpc-status", "invalid"}},
+        },
+    }));
+
+TEST_P(BadResponseGrpcHealthCheckerImplTest, GrpcBadResponse) {
+  setupHC();
+  cluster_->prioritySet().getMockHostSet(0)->hosts_ = {
+      makeTestHost(cluster_->info_, "tcp://127.0.0.1:80")};
+  expectSessionCreate();
+  expectStreamCreate(0);
+  EXPECT_CALL(*test_sessions_[0]->timeout_timer_, enableTimer(_));
+  health_checker_->start();
```

Comment:
I think there is also some common setup code that could be factored out.

Reply:
done